tests/end_to_end/helpers.h

```diff
@@ -89,19 +89,37 @@ typedef class _single_instance_hook : public _hook_helper
                 client_attach_callback,
                 client_detach_callback) == EBPF_SUCCESS);
     }
-    ~_single_instance_hook() { ebpf_provider_unload(provider); }
+    ~_single_instance_hook()
+    {
+        detach();
+        ebpf_provider_unload(provider);
+    }
 
     uint32_t
     attach(ebpf_handle_t program_handle)
     {
         return ebpf_api_link_program(program_handle, attach_type, &link_handle);
     }
 
+    uint32_t
+    attach(bpf_program* program)
+    {
+        return ebpf_program_attach(program, &attach_type, nullptr, 0, &link_object);
+    }
+
     void
     detach()
     {
-        ebpf_api_unlink_program(link_handle);
-        ebpf_api_close_handle(link_handle);
+        if (link_handle != ebpf_handle_invalid) {
```

Comment:
Is this because this is a test hook? I dont think for real link, we need to do these?

Reply:
I don't understand your question.

Reply:
The test code can either accept:
1) program handle
2) program object

on detach, we need to cleanup depending on how it was attached.